FastSimulation/Layer/src/LayerFactory.cc

```diff
@@ -0,0 +1,267 @@
+    // ,and check order.
+    for(unsigned index = 1;index < limits.size();index++)
+    {
+	if(limits[index] < limits[index-1])
+	{
+	    std::string cfgString;
+	    cfg.allToString(cfgString);
+	    throw cms::Exception("fastsim::LayerFactory") 
+		<< "limits must be provided in increasing order. error in:\n"
+		<< cfgString;
+	}
+    }
+    // Get thickness values
+    const std::vector<double> & thickness = cfg.getUntrackedParameter<std::vector<double> >("thickness");
+    // , and check compatibility with limits
+    if(limits.size() < 2 || thickness.size() != limits.size() - 1)
+    {
+	std::string cfgString;
+	cfg.allToString(cfgString);
+	throw cms::Exception("fastim::LayerFactory") 
+	    << "layer thickness and limits not configured properly! error in:"
+	    << cfgString;
+    }
+    // create the histogram
+    layer->thicknessHist_.reset(new TH1F("h", "h", limits.size()-1, &limits[0]));
+    layer->thicknessHist_->SetDirectory(0);
+    for(unsigned i = 1; i < limits.size(); ++i){
+	layer->thicknessHist_->SetBinContent(i, thickness[i-1]);
+    }
+    
+    // -----------------------------
+    // nuclear interaction thickness factor
+    // -----------------------------
+
+    layer->nuclearInteractionThicknessFactor_ = cfg.getUntrackedParameter<double>("nuclearInteractionThicknessFactor",1.);
+
+    // -----------------------------
+    // magnetic field
+    // -----------------------------
+    
+    layer->magneticFieldHist_.reset(new TH1F("h", "h", 100, 0., isForward ? magneticFieldHistMaxR_ : magneticFieldHistMaxZ_));
+    layer->magneticFieldHist_->SetDirectory(0);
+    for(int i = 1; i <= 101; i++)
+    {
+	GlobalPoint point = isForward ? 
+	    GlobalPoint(layer->magneticFieldHist_->GetXaxis()->GetBinCenter(i), 0.,position)
+	    : GlobalPoint(position, 0.,layer->magneticFieldHist_->GetXaxis()->GetBinCenter(i));
+	layer->magneticFieldHist_->SetBinContent(i, magneticField_->inTesla(point).z());
+    }
+    
+    // -----------------------------
+    // list of interaction models
+    // -----------------------------
+
+    std::vector<std::string> interactionModelLabels = cfg.getUntrackedParameter<std::vector<std::string> >("interactionModels");
+    for(const auto & label : interactionModelLabels)
+    {
+    	std::map<std::string,fastsim::InteractionModel *>::const_iterator interactionModel = interactionModelMap_->find(label);
+    	if(interactionModel == interactionModelMap_->end())
+    	{
+    	    throw cms::Exception("fastsim::LayerFactory") << "unknown interaction model '" << label << "'";
+    	}
+    	layer->interactionModels_.push_back(interactionModel->second);
+    }
+
+    // -----------------------------
+    // and return the layer!
+    // -----------------------------
+
+    return std::move(layer);
+} 
+
+const DetLayer *
+fastsim::LayerFactory::getDetLayer(const std::string & detLayerName, const GeometricSearchTracker & geometricSearchTracker) const
```

Comment:
I need to return a pointer to a DetLayer. I'm not quite sure how to do this with the function you are pointing me to.